solr/core/src/java/org/apache/solr/handler/component/UBIComponent.java

```diff
@@ -0,0 +1,424 @@
+
+    // not sure why but sometimes we get it twoice...  how can a response have the
+    // the same component run twice?
+    if (rb.rsp.getValues().get("ubi") != null) {
+      return;
+    }
+    SolrParams params = rb.req.getParams();
+
+    SolrIndexSearcher searcher = rb.req.getSearcher();
+
+    String queryId = params.get(QUERY_ID);
+    UBIQuery ubiQuery = new UBIQuery(queryId);
+
+    ubiQuery.setUserQuery(params.get(USER_QUERY));
+    ubiQuery.setApplication(params.get(APPLICATION));
+    if (ubiQuery.getApplication() == null) {
+      ubiQuery.setApplication(
+          rb.isDistrib
+              ? rb.req.getCloudDescriptor().getCollectionName()
+              : searcher.getCore().getName());
+    }
+
+    String queryAttributes = params.get(QUERY_ATTRIBUTES);
+
+    if (queryAttributes != null && queryAttributes.toString().startsWith("{")) {
+      // Look up the original nested JSON format, typically passed in
+      // via the JSON formatted query.
+      @SuppressWarnings("rawtypes")
+      Map jsonProperties = rb.req.getJSON();
+      if (jsonProperties.containsKey("params")) {
+        @SuppressWarnings("rawtypes")
+        Map paramsProperties = (Map) jsonProperties.get("params");
+        if (paramsProperties.containsKey(QUERY_ATTRIBUTES)) {
+          @SuppressWarnings("rawtypes")
+          Map queryAttributesAsMap = (Map) paramsProperties.get(QUERY_ATTRIBUTES);
+          ubiQuery.setQueryAttributes(queryAttributesAsMap);
+        }
+      }
+    }
+
+    ResultContext rc = (ResultContext) rb.rsp.getResponse();
+    DocList docs = rc.getDocList();
+    // DocList docs = rb.getResults().docList;
+
+    String docIds = extractDocIds(docs, searcher);
+    ubiQuery.setDocIds(docIds);
+
+    addUserBehaviorInsightsToResponse(ubiQuery, rb);
+    recordQuery(ubiQuery);
+  }
+
+  private void recordQuery(UBIQuery ubiQuery) throws IOException {
+    TupleStream stream;
+
+    stream = constructStream(streamFactory, streamExpression);
+
+    streamContext.put("ubi-query", ubiQuery);
+    stream.setStreamContext(streamContext);
+
+    getTuple(stream);
+  }
+
+  private void addUserBehaviorInsightsToResponse(UBIQuery ubiQuery, ResponseBuilder rb) {
+    SimpleOrderedMap<String> ubiResponseInfo = new SimpleOrderedMap<>();
+
+    ubiResponseInfo.add(QUERY_ID, ubiQuery.getQueryId());
+    rb.rsp.add("ubi", ubiResponseInfo);
+  }
+
+  protected String extractDocIds(DocList dl, SolrIndexSearcher searcher) throws IOException {
+    IndexSchema schema = searcher.getSchema();
+
+    if (schema.getUniqueKeyField() == null) {
+      log.error("Can't track documents for query without unique field.");
+      return "";
+    }
+    StringBuilder sb = new StringBuilder();
+
+    Set<String> fields = Collections.singleton(schema.getUniqueKeyField().getName());
+    for (DocIterator iter = dl.iterator(); iter.hasNext(); ) {
+      sb.append(schema.printableUniqueKey(searcher.getDocFetcher().doc(iter.nextDoc(), fields)))
```

Comment:
what about commas in id values? Isn't it safer to use json array as a convention for this field?

Reply:
humm...    maybe?   Is that at all ever done?   So, most folks are going to take this data, and load it into a pandas dataframe or soemthing like that...    Now, since the doc is already JSON..   maybe it's okay to make this a array of JSON as well?    Since you already have to parse some JSON to do anything with it......